File: C++/LinkedList/LinkedListNthFromEnd.cpp
/*
	@author
   	Aakash Verma
	
 	Given a Linked List, find the nth node from the end of a linked list.

 	Output:

	Original List is: 2 3 4 5 6 
	4


	Compile using:  g++ filename.cpp
	Run using: ./a.out filename.cpp

*/


#include <iostream>
using namespace std;

/* creating a node structure in C/C++ */
struct Node {
   int data;
   struct Node *next;
};

/* defining head as null as there is no node in the list initially. */
struct Node* head = NULL;

void push(int data) {
	Node *temp;
	temp = head;

	/* creating a node and assigning data to it and make its next as null */
	struct Node* new_node = (struct Node*) malloc(sizeof(struct Node));
	new_node->data = data;
   	new_node->next = NULL;

   	/* check if it's empty make new_node as head */
	if(head == NULL) {
   		head = new_node;
	}
	/* otherwise move upto the last and then connect last node with the new_node */
	else { 
		while(temp->next != NULL) {
			temp = temp->next;
		}
   		temp->next = new_node;
	}
}

/* Utility function to find the nth Node from the end. */
void nthNodeFromEnd(int nthNode) {
	Node *prev, *curr;
	prev = head;
	curr = head;

	if(head == NULL) {
		cout << "The list doesn't exist." << endl;
		return;
	}

	/* curr pointer is being moved ahead n times. */
	for(int i = 0; i < nthNode; i++) {
		if(curr != NULL) {
			curr = curr->next;	
		}
		else {
			cout << "Enough nodes are not present in the linked list." << endl;
			return;
		}
	}

	// Now the difference between prev and curr pointer is n

	/* now we'll move prev and curr pointer both until curr becomes null and finally prev will be at n from last */
	while(curr != NULL) {
		curr = curr->next;
		prev = prev->next;
	}

	/* finally printing data of nth node from last */
	cout << prev->data << endl;

}

/* for printing a ist */
void printList() {

   	struct Node* ptr;
   	ptr = head;
   	
   	while (ptr != NULL) {
      	cout<< ptr->data <<" ";  // use printf() for C
      	ptr = ptr->next;
   	}
   	cout<<endl;
}

int main() {

	push(2);
	push(3);
	push(4);
	push(5);
	push(6);
	cout<<"Original List is: ";
	printList();
	nthNodeFromEnd(3);

	return 0;

}

File: Python/LinkedList/LinkedListNthFromEnd.py
#
#
#
#   @author
#   Aakash Verma
#	
# 	Nth Node from the Last of a Linked List
#
#	Output: 
#
#	Original List is:  3 4 5 6 7 9 
#	6
#


# Below is the structute of a node which is used to create a new node every time.
class Node: 

	def __init__(self, data): 
		self.data = data
		self.next = None # None is nothing but null 

# Creating a class for implementing the code for nth Node from end in a Linked List
class LinkedList: 
	
	# Whenever I'll create the object of a LinkedList class head will be pointing to null initially
	def __init__(self): 
		self.head = None


	# Inserting at the end of a Linked List (append function)
	def append(self, key): 

		h = self.head

		if h is None:
			new_node = Node(key)
			self.head = new_node
		else:
			while h.next != None:
				h = h.next
			new_node = Node(key)
			h.next = new_node

	# Utility function to print the linked list 
	def printList(self): 
		h = self.head 
		while (h): 
			print (h.data, end = " ") 
			h = h.next
		print()

	# nth Node from end
	def nthNodeFromEnd(self, nthNode):
		prev = self.head
		curr = self.head
		if self.head is None:
			print("The list doesn't exist.")
			return

		# curr pointer is being moved ahead n times.
		for i in range(nthNode):
			if curr is not None:
				curr = curr.next
			else:
				print("Enough nodes are not present in the linked list.")
				return

		# Now the difference between prev and curr pointer is n

		# now we'll move prev and curr pointer both until curr becomes null and finally prev will be at n from last
		while curr is not None:
			curr = curr.next
			prev = prev.next

		# finally printing data of nth node from last
		print(prev.data)


# Code execution starts here 
if __name__=='__main__': 

	myList = LinkedList() 
	myList.append(3)
	myList.append(4)
	myList.append(5)
	myList.append(6)
	myList.append(7)
	myList.append(9) 
	print("Original List is: ", end = " ")
	myList.printList()
	myList.nthNodeFromEnd(3)

